### ISSUE TITLE
fixes the environment.yml for Albedo (AWI) - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1789]
In GitLab by @mandresm on Jan 11, 2023, 20:04

_Merges fix/conda_environment -> master_

This merge fixes 2 problems we are having at AWI to install Autosubmit in our machine via conda:
1. The python minor version of the command `conda env create -f environment.yml -n autosubmit python-3.7` is ignored.
2. Even when the correct version of python is install in the conda environment it still tries to use the system's default one. That's solved by adding `ipython` to the conda environment.

@dbeltrankyl, I'm not sure if `master` is the correct branch to merge-request in this case.

**Comment by kinow** [comment]
In GitLab by @mandresm on Jan 11, 2023, 20:04

requested review from @dbeltrankyl

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 11, 2023, 20:23

@kinow can you have a look too?

**Comment by kinow** [comment]
@mandresm , I used `master`, applying first the `python=3.7`, and then:

- `conda env remove -n autosubmit` to delete my env
- `conda env create -f environment.yml -n autosubmit` to create it again (the `environment.yml` now contains `python=3.7`)
- `conda activate autosubmit`

```bash
(autosubmit) kinow@ranma:~/Development/python/workspace/autosubmit$ python
Python 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:21) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

Could you test again if `ipython` is needed, please?

Cheers

**Comment by kinow** [comment]
In GitLab by @mandresm on Jan 12, 2023, 10:13

Commented on [environment.yml line 7](https://github.com/BSC-ES/autosubmit/compare/3a85aae0c48733d2b9ab3567aefec69d75c3f49c..1fec54f52e0da57aa34a13704fe07900a7ce9f28#diff-b7cb75ca59993a879e29bed8cd8469b8R7)



I can confirm that without ipython now works. I'll commit that now.

**Comment by kinow** [comment]
In GitLab by @mandresm on Jan 12, 2023, 10:16

Commented on [environment.yml line 7](https://github.com/BSC-ES/autosubmit/compare/3a85aae0c48733d2b9ab3567aefec69d75c3f49c..1fec54f52e0da57aa34a13704fe07900a7ce9f28#diff-b7cb75ca59993a879e29bed8cd8469b8R7)



changed this line in [version 2 of the diff](/gitlab/es/autosubmit/-/merge_requests/301/diffs?diff_id=11731&start_sha=1fec54f52e0da57aa34a13704fe07900a7ce9f28#54c391de2500e0856c23737daf64f9d931569110_7_7)

**Comment by kinow** [comment]
approved this merge request

**Comment by kinow** [comment]
mentioned in commit 2136a8a9ad3783badc71c509b6ebf657ec1c475d

**Comment by kinow** [comment]
In GitLab by @mandresm on Jan 12, 2023, 10:28

Thanks @kinow for this very fast merge! For the DestinE workflow, should we switch to the `master` branch or should we stick to the tag `v4.0.0b`? In that later case we would need to move the tag to the current commit?

**Comment by kinow** [comment]
Not a problem, thanks for the merge request :slight_smile: 

>For the DestinE workflow, should we switch to the `master` branch or should we stick to the tag `v4.0.0b`? In that later case we would need to move the tag to the current commit?

Not sure, I'm using the versions from PyPI for documents, but for some tests I use the version from `master` too.

@dbeltrankyl , @mcastril ?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jan 12, 2023, 11:01

It is better to use the 4.0.0b tag as the master can be more unstable.

4.0.0b will be getting updates until it ends its beta phase. So the commit will change (and also the autosubmitconfigparser dependency may change so is better to also update the dependencies) 

Pip version will also get updates

**Comment by kinow** [comment]
In GitLab by @sebastianbeyer on Jan 12, 2023, 11:49

Hey, maybe not the right place to discuss this, but I would suggest not to change the code that responds to a tag. That way if someone reports a bug saying it is on version v4.0.0b you can not be sure what exact version/commit that responds to. What do you think about making a new tag for every "release" according to [semantic versioning](https://semver.org/)?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 12, 2023, 12:52

Hi @sebastianbeyer, I see your point. We could use the third (Z) digit to differentiate minor updates. The reality is that versions tagged as (b) are development/unstable versions that take months to be promoted to a major (Y) release. And usually, dozens of updates are needed. This is because we don't want to wait too much to fix bugs and publish the bug fixes (grouping different fixes together in minor releases). This is also because users consistently access the code through the tag (which has a corresponding module in our system).

Anyway, we can discuss internally if we can adapt our current protocol and have a `master` module. Then, create a tag from time to time.

**Comment by kinow** [comment]
>What do you think about making a new tag for every "release" according to [semantic versioning](https://semver.org/)?

That sounds like a good idea to me. I think most external users would expect tags to be immutable,

- https://git-scm.com/docs/git-tag#_discussion

We could have a 4.0.0b branch and create multiple releases from this branch. Each release would receive a [PEP-440](https://peps.python.org/pep-0440/#summary-of-permitted-suffixes-and-relative-ordering) version ID like 4.0.0b1, 4.0.0b2, 4.0.0b3. And each of these beta releases would result in a tag with the same suffix b1, b2, and so on, that are immutable.

**Comment by kinow** [comment]
In GitLab by @mandresm on Jan 12, 2023, 13:36

> We could have a 4.0.0b branch and create multiple releases from this branch. Each release would receive a [PEP-440](https://peps.python.org/pep-0440/#summary-of-permitted-suffixes-and-relative-ordering) version ID like 4.0.0b1, 4.0.0b2, 4.0.0b3. And each of these beta releases would result in a tag with the same suffix b1, b2, and so on, that are immutable.

I was gonna suggest precisely that: we have a branch for this. We can treat it as the tag has been used right now, a more stable branch for general purpose testing, that can be periodically updated with the master. That would make things easier for the WF development team at DestinE in terms of getting and sharing the necessary updates, merge-requesting, etc... Tagging as @sebastianbeyer and @kinow would be very nice too :)

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 12, 2023, 14:21

I also think that it will be better organized with a branch. However, I think it doesn't solve the correspondence between releases and our internal modules that I was commenting on above.

I agree that it would be the right way to go, but we need to discuss the procedure a little bit because it adds some overhead to the current flow.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jan 12, 2023, 16:26

We had an internal meeting and we agreed to:

* Git: We will keep master branch. The changes will be first done on a dev branch related to the issue and when it is ready, we will merge it to master branch and create a new tag(likely we will bundle more than one change at once to avoid spam of versions). 4.0.0b01,4.0.0b02 ... This will be the procedure from now. 

Only when the beta is released we will create the 4.0.0-branch 

* Version file: Will be always version 4.0.0b without the additional number.

* (bsc only) Our modules: Will keep the same name, I'll just adapt the script to point to the newest tag. 

*  Pip: We will point to the 4.1.0b01,4.1.0b02.. we can't do it for 4.0.0 since the current version is already 4.0.7.
